File: tape/sdk/go/embedded/decorators.go
package embedded

// decorators.go — Go's answer to Python's `@effect` and `@outbox_tool`.
// We don't have decorators; we have higher-order constructors. The
// returned value (`*EffectTool` / `*OutboxTool`) carries the metadata
// the (future) ADK-Go plugin would read at call time.
//
// Construction-time refusal is the load-bearing property: if the
// caller asks for the NON-IDEMPOTENT + OUTBOX shape without a way to
// recover, we fail HERE — before any agent ever runs the tool.

import (
	"errors"
	"fmt"
)

// ToolFn — the user-supplied tool body. Args are JSON-coercible
// (typically the model's parsed function-call arguments). The first
// return value is the tool result; the second is the journaled error.
type ToolFn func(args map[string]any) (any, error)

// CustomKeyFn — derive an idempotency key from the tool args (overrides
// the default `<invocation>/decision-<n>/<tool>/<call>` shape).
type CustomKeyFn func(args map[string]any) string

// BusinessKeyFn — derive the cross-run dedup key for the OUTBOX path.
type BusinessKeyFn func(args map[string]any) (string, error)

// EffectMeta — metadata stamped on a tool by `Effect` / `OutboxTool`.
// Read it back via `MetaOf(tool)`.
type EffectMeta struct {
	Semantics         string
	DispatchMode      string
	BusinessKeyFn     BusinessKeyFn
	BusinessKeyStatic string
	Connector         string
	Compensate        string
	CustomKeyFn       CustomKeyFn
}

// ResolveBusinessKey — apply BusinessKeyStatic if set, else BusinessKeyFn.
// Returns ("", nil) when neither is configured.
func (m EffectMeta) ResolveBusinessKey(args map[string]any) (string, error) {
	if m.BusinessKeyStatic != "" {
		return m.BusinessKeyStatic, nil
	}
	if m.BusinessKeyFn != nil {
		return m.BusinessKeyFn(args)
	}
	return "", nil
}

// EffectTool — the value returned by `Effect(...)`. Carries the tool
// function plus its metadata; pass to your plugin / runner.
type EffectTool struct {
	Fn   ToolFn
	Meta EffectMeta
}

// Call — invoke the wrapped function. Convenience for callers that just
// want to run the body directly (replay / testing).
func (t *EffectTool) Call(args map[string]any) (any, error) { return t.Fn(args) }

// EffectOpts — options for `Effect`.
type EffectOpts struct {
	CustomKey CustomKeyFn
}

// Effect — wrap `fn` as an IDEMPOTENT + INLINE tool. The plugin records
// the intent before the call, the result after, and short-circuits on
// replay. The body MUST be safe to call multiple times (the upstream
// dedupes via the request's idempotency key, or the body is a no-op on
// repeat).
func Effect(fn ToolFn, opts ...EffectOpts) *EffectTool {
	var o EffectOpts
	if len(opts) > 0 {
		o = opts[0]
	}
	return &EffectTool{
		Fn: fn,
		Meta: EffectMeta{
			Semantics:    EffectSemanticsIdempotent,
			DispatchMode: EffectDispatchInline,
			CustomKeyFn:  o.CustomKey,
		},
	}
}

// OutboxToolOpts — options for `OutboxTool`. `Connector` is required;
// exactly one of `BusinessKeyFn` / `BusinessKeyStatic` must be set.
type OutboxToolOpts struct {
	// BusinessKeyFn — derive the cross-run dedup key from the tool
	// args. Mutually exclusive with BusinessKeyStatic.
	BusinessKeyFn BusinessKeyFn

	// BusinessKeyStatic — for cases where the business key is constant
	// (rare). Mutually exclusive with BusinessKeyFn.
	BusinessKeyStatic string

	// Connector — the registry name the outbox dispatcher resolves at
	// runtime. Must match the `Name()` on a registered Connector.
	Connector string

	// Compensate — the obligation `kind` to register on duplicate
	// observation. Mirrors `compensate_on_duplicate_kind` in the proto.
	Compensate string

	// CustomKey — overrides the default idempotency-key derivation.
	CustomKey CustomKeyFn
}

// ErrOutboxToolConfig — returned by `OutboxTool` when the configuration
// is unsafe. Matches `Python tape_adk.decorators.outbox_tool`'s
// construction-time `ValueError`s.
var ErrOutboxToolConfig = errors.New("OutboxTool: invalid configuration")

// OutboxTool — wrap `fn` as a NON_IDEMPOTENT + OUTBOX tool. The plugin
// never lets the body run inline; it journals an intent and the outbox
// dispatcher (a separate reactor) makes the upstream call.
//
// Construction-time refusal — returns an error wrapping
// `ErrOutboxToolConfig` if:
//
//   - `Connector` is empty;
//   - both `BusinessKeyFn` and `BusinessKeyStatic` are zero;
//   - both `BusinessKeyFn` and `BusinessKeyStatic` are set
//     (ambiguous — pick one).
//
// The bug never makes it past compile-time-adjacent validation.
func OutboxTool(fn ToolFn, opts OutboxToolOpts) (*EffectTool, error) {
	if opts.Connector == "" {
		return nil, fmt.Errorf("%w: `Connector` is required — the outbox dispatcher "+
			"needs to know which connector to dispatch through", ErrOutboxToolConfig)
	}
	if opts.BusinessKeyFn == nil && opts.BusinessKeyStatic == "" {
		return nil, fmt.Errorf("%w: `BusinessKeyFn` or `BusinessKeyStatic` is required — "+
			"non-idempotent operations must declare the key the upstream uses to dedupe",
			ErrOutboxToolConfig)
	}
	if opts.BusinessKeyFn != nil && opts.BusinessKeyStatic != "" {
		return nil, fmt.Errorf("%w: set exactly one of `BusinessKeyFn` or `BusinessKeyStatic`",
			ErrOutboxToolConfig)
	}
	return &EffectTool{
		Fn: fn,
		Meta: EffectMeta{
			Semantics:         EffectSemanticsNonIdempotent,
			DispatchMode:      EffectDispatchOutbox,
			BusinessKeyFn:     opts.BusinessKeyFn,
			BusinessKeyStatic: opts.BusinessKeyStatic,
			Connector:         opts.Connector,
			Compensate:        opts.Compensate,
			CustomKeyFn:       opts.CustomKey,
		},
	}, nil
}

// MustOutboxTool — panic-on-misconfiguration helper for `init()` blocks
// where an unsafe configuration should crash the process immediately.
func MustOutboxTool(fn ToolFn, opts OutboxToolOpts) *EffectTool {
	t, err := OutboxTool(fn, opts)
	if err != nil {
		panic(err)
	}
	return t
}

// MetaOf — read the metadata off an `*EffectTool`. Returns the zero
// value if `t` is nil.
func MetaOf(t *EffectTool) EffectMeta {
	if t == nil {
		return EffectMeta{}
	}
	return t.Meta
}

File: tape/sdk/typescript/src/embedded/decorators.ts
// `effect` and `outboxTool` — attach Tape metadata to a tool function so a
// hosting framework (or a user's own dispatcher) can recognise it at call
// time and journal accordingly. Mirrors `tape_adk/decorators.py`.
//
// Since TypeScript doesn't have true decorators on plain functions (the
// `@` syntax requires class members), we expose these as higher-order
// wrappers: `effect(fn)` and `outboxTool({…}, fn)`.
//
// Construction-time refusal:
// * `outboxTool` requires `businessKey` AND `connector`. Missing either
//   throws `Error` *immediately* — the bug never makes it past import.

import { EffectDispatchMode, EffectSemantics } from './service.ts';

/** Metadata stamped on a wrapped function. */
export interface EffectMeta {
  semantics: string;
  dispatchMode: string;
  businessKeyFn?: (args: Record<string, unknown>) => string;
  businessKeyStatic?: string;
  connector?: string;
  compensate?: string;
  customKeyFn?: (args: Record<string, unknown>) => string;
}

// We stash the metadata on the function under a Symbol so it doesn't
// collide with user properties.
const TAPE_META = Symbol.for('tape.embedded.effect_meta');

type FnWithMeta<F extends (...a: never[]) => unknown> = F & {
  [TAPE_META]?: EffectMeta;
};

/** Read Tape metadata off a wrapped function. Returns undefined if not Tape-tracked. */
export function metaOf<F extends (...a: never[]) => unknown>(fn: F): EffectMeta | undefined {
  return (fn as FnWithMeta<F>)[TAPE_META];
}

/**
 * Mark a tool as idempotent + inline-journaled. The host records the intent
 * before the call, the result after, and short-circuits on replay.
 *
 * The tool body MUST be safe to call multiple times — if the agent crashes
 * between intent and result, the re-drive will call the body again. The
 * upstream is expected to dedupe via its own idempotency key, or the body
 * itself must be a no-op on repeat.
 *
 * Usage:
 *   const myTool = effect(async (args) => { ... });
 *   const myTool = effect(async (args) => { ... }, { customKey: a => a.id });
 */
export function effect<F extends (...a: never[]) => unknown>(
  fn: F,
  opts: { customKey?: (args: Record<string, unknown>) => string } = {},
): F {
  const meta: EffectMeta = {
    semantics: EffectSemantics.IDEMPOTENT,
    dispatchMode: EffectDispatchMode.INLINE,
    customKeyFn: opts.customKey,
  };
  (fn as FnWithMeta<F>)[TAPE_META] = meta;
  return fn;
}

/**
 * Mark a tool as NON-IDEMPOTENT — its dispatch lives in the outbox.
 *
 * The agent calls the tool conceptually; the host intercepts and journals
 * an intent. The actual upstream call is made later by the outbox dispatcher
 * reactor, against the connector named here.
 *
 * Required (refused at construction time if missing):
 * * `businessKey` — string or callable that derives the upstream's dedupe key
 *   from the tool's args. The (connector, business_key) tuple is UNIQUE
 *   across the journal — no two effects can share it for the same connector.
 * * `connector` — registry name the outbox dispatcher resolves at runtime.
 *
 * Optional:
 * * `compensate` — obligation `kind` to register on duplicate observation.
 * * `customKey` — override the derived idempotency_key.
 *
 * Construction-time refusal: omitting `businessKey` or `connector` throws
 * `Error` here — the bug never makes it past `import`.
 */
export function outboxTool<F extends (...a: never[]) => unknown>(
  opts: {
    businessKey: string | ((args: Record<string, unknown>) => string);
    connector: string;
    compensate?: string;
    customKey?: (args: Record<string, unknown>) => string;
  },
  fn: F,
): F {
  if (!opts.connector) {
    throw new Error(
      'outboxTool: `connector` is required — the outbox dispatcher needs '
      + 'to know which connector to dispatch through.');
  }
  if (opts.businessKey === undefined || opts.businessKey === null) {
    throw new Error(
      'outboxTool: `businessKey` is required — non-idempotent operations '
      + 'must declare the key the upstream uses to dedupe. Pass a string '
      + 'OR a callable that derives it from the tool args.');
  }
  const meta: EffectMeta = {
    semantics: EffectSemantics.NON_IDEMPOTENT,
    dispatchMode: EffectDispatchMode.OUTBOX,
    connector: opts.connector,
    compensate: opts.compensate,
    customKeyFn: opts.customKey,
  };
  if (typeof opts.businessKey === 'string') {
    meta.businessKeyStatic = opts.businessKey;
  } else {
    meta.businessKeyFn = opts.businessKey;
  }
  (fn as FnWithMeta<F>)[TAPE_META] = meta;
  return fn;
}
